### ISSUE TITLE
problem with cumsum

### ISSUE THREAD

**Issue by my-little-repository** [JuliaLang/julia#13244]
There seems to be a type problem with the `cumsum` function when passed an union type. 

     julia> f(x) = (x == 0 ? 0 : 0.5) 
     f (generic function with 1 method)
     
     julia> cumsum([f(x) for x = 0:1])     # Array{Union{Float64,Int64},1}
     ERROR: InexactError()
      in convert at int.jl:209
      in cumsum_pairwise! at arraymath.jl:431
      in cumsum! at arraymath.jl:445
      in anonymous at no file
     
     julia> sum([f(x) for x = 0:1])        # no problem here
     0.5
     
The error does not appear with the function `f(x) = (x == 0 ? 0 : 1.0)`. An `Int` array is returned.

**Comment by aaneto** [comment]
Well, I'm no core developer but,
I've looked on int.jl:209 and found this:
<pre><code>
for to in (Int8, Int16, Int32, Int64)
    @eval begin
        convert(::Type{$to}, x::Float32) = box($to,checked_fptosi($to,unbox(Float32,x)))
        convert(::Type{$to}, x::Float64) = box($to,checked_fptosi($to,unbox(Float64,x))) <- line 209
    end
end
</code></pre>
Well, julia is on the case of convert(Int64, x::Float64) setting the function to <code>box(Int64,checked_fptosi(Int64,unbox(Float64,x)))</code> which I believe is what is happening, since cumsum tries to sum 0 to 0.5, then the convert function is failing:

<pre><code>
convert(Int64, 2.3)
ERROR: InexactError()
 in convert at ./int.jl:209
</code>
</pre>

For what it looks like, unbox is returning a float, and it shouldn't.

**Comment by simonster** [comment]
It's intended behavior that converting a non-integer floating point number to an integer throws an InexactError. I'm pretty sure that the problem is that `cumsum` [decides the type of the output based on the first element of the input](https://github.com/JuliaLang/julia/blob/f172f4e95d34c50a4885a5238bdc4ebe54224dac/base/arraymath.jl#L420).

**Comment by tkelman** [comment]
The first element thing isn't right for heterogeneous arrays, we should maybe do some better specialization there to only do that when the element type is concrete?

There was an argument a while back in https://github.com/JuliaLang/julia/pull/9665 about whether cumsum should promote to larger types like sum does to make it less likely to overflow, I personally don't think that's right since it returns an array rather than a scalar and the storage could grow pretty substantially.

**Comment by simonster** [comment]
Yes, we should do something better with heterogeneous arrays, both here and with `mapreducedim`. The right solution is probably to do what `map` does and start with an array that is typed according to the first element and then widen its type as necessary.

**Comment by my-little-repository** [comment]
When it comes to array, `cumsum` sends `[x_1,....x_n]` to `[x_1, x_1+x_2, ...., x_1+x_2+..+x_n]`. Since `+` overflows instead of converting to big for int and floats, my guess is that the type of the array returned by `cumsum` should be equal to the type of the array passed as an argument.  `+` can be overloaded though.

I must say that the following line, referenced above,  is a mystery to me. Why adding `v[1]` to itself and taking its type?

      _cumsum_type(v) = typeof(v[1]+v[1])

**Comment by KristofferC** [comment]
I guess all types doesn't have to be closed during  addition.

**Comment by andreasnoack** [comment]
E.g. Bool

Den onsdag den 23. september 2015 skrev Kristoffer Carlsson <
notifications@github.com>:

**Comment by hayd** [comment]
Maybe cumsum (and friends) could optionally take a first argument of the type you want to reduce on.

**Comment by tkelman** [comment]
You can do `cumsum!` into an array with element type of your choosing

**Comment by hayd** [comment]
Interestingly I think this actually boils down to:
```
julia> typeof([f(x) for x = 0:1])
Array{Union{Float64,Int64},1}

julia> Base._cumsum_type([f(x) for x = 0:1])  # IMO expect Float64
Int64

julia> typeof(zero(Union{Float64,Int64}))  # stems from
Int64
```
which is due to [this definition](https://github.com/JuliaLang/julia/blob/7c8153b585e85f7b5b3b8a799530a479f843cbd2/base/essentials.jl#L58) (since zero just converts 0):
```
convert{T}(::Type{T}, x::T) = x
```

The following definition fixes it but with ambiguity warnings galore:
This should probably be the

**Comment by stevengj** [comment]
I think this is a bug in `_cumsum_type`; should be fixed in #18364.

**Comment by stevengj** [comment]
Closed by #18364.
